components/region_cache_memory_engine/src/engine.rs

```diff
@@ -246,71 +259,61 @@ impl RangeCacheMemoryEngine {
     // It handles the pending range and check whether to buffer write for this
     // range.
     //
-    // Return `(range_in_cache, pending_range_in_loading)`, see comments in
-    // `RangeCacheWriteBatch` for the detail of them.
-    //
-    // In addition, the region with range equals to the range in the `pending_range`
-    // may have been splited, and we should split the range accrodingly.
+    // Return `(bool, bool)`, the first indicates the range is in cache, the second
+    // indicates the range is in loading, see comments in `RangeCacheWriteBatch`
+    // for the detail of them.
     pub(crate) fn prepare_for_apply(&self, range: &CacheRange) -> (bool, bool) {
         let core = self.core.upgradable_read();
         let range_manager = core.range_manager();
-        let mut pending_range_in_loading = range_manager.pending_ranges_in_loading_contains(range);
-        let range_in_cache = range_manager.contains_range(range);
-        if range_in_cache || pending_range_in_loading {
-            return (range_in_cache, pending_range_in_loading);
-        }
-
-        // check whether the range is in pending_range and also split it if the range
-        // has been splitted
-        let mut index = None;
-        let mut left_splitted_range = None;
-        let mut right_splitted_range = None;
-        for (i, r) in range_manager.pending_ranges.iter().enumerate() {
-            if r == range {
-                index = Some(i);
-                break;
-            } else if r.contains_range(range) {
-                index = Some(i);
-                // It means the loading region has been splitted. We split the
-                // range accordingly.
-                (left_splitted_range, right_splitted_range) = r.split_off(range);
-                break;
-            } else if range.contains_range(r) {
-                // todo: it means merge happens
-                unimplemented!()
+        if range_manager.pending_ranges_in_loading_contains(range) {
+            return (false, true);
```

Comment:
It seems that range_in_cache and pendign_range_in_loading is exclusive (they cannot be true at the same time).
It seems to me it's more clearer to have an enum instead. 
enum RangeCacheStatus {
      NotInCache,
      Cached,
      Loading, 
}

Reply:
Sounds greate. Done.